torch/_inductor/codegen/triton.py

```diff
@@ -1171,6 +1172,49 @@ def codegen_body(self):
         self.stores.clear()
         self.suffix.clear()
 
+    def codegen_kernel_benchmark(self):
+        result = IndentedBuffer()
+
+        result.writelines(["", "", 'if __name__ == "__main__":'])
+        grid = []
+        with result.indent():
+            result.writeline("from torch._C import _cuda_getCurrentRawStream as get_cuda_stream")
+            result.writeline("from torch._dynamo.testing import rand_strided")
+            result.writeline("from torch._inductor.utils import print_performance")
+            result.writeline("import torch")
+            result.writeline("from torch._inductor.triton_ops.autotune import grid")
+            result.writeline("")
+
+            argdefs, call_args, signature = self.args.python_argdefs()
+
+            for arg_name in call_args:
+                buf = V.graph.get_buffer(arg_name)
+                if buf:
+                    result.writeline(
+                        f"{arg_name} = rand_strided({tuple(buf.get_size())}, {tuple(buf.get_stride())}, device='{buf.get_device()}', dtype={buf.get_dtype()})"
+                    )
+
+            for tree in self.range_trees:
+                expr = pexpr(tree.numel)
+                if tree.prefix != "r" or self.inside_reduction:
+                    call_args.append(expr)
+                if tree.prefix != "r":
+                    grid.append(expr)
+
+            call_args = ", ".join(call_args)
+            index = V.graph.scheduler.current_device.index
+            stream_name = f"stream{index}"
+            result.writeline(f"{stream_name} = get_cuda_stream({index})")
+
+            # print_performance assert non-None return value, but a triton kernel
+            # indeed returns None. Override the return value to True to handle this.
+            result.writeline(
+                f"print_performance(lambda: triton_.run({call_args}, grid=grid({', '.join(grid)}), stream={stream_name}) or True)"
```

Comment:
@chillee I'm not sure we want `print_performance` here, should `TORCHINDUCTOR_PROFILE` be enough?

Reply:
AFAICT, `TORCHINDUCTOR_PROFILE` collects latency and bandwidth information for each triton kernel being called. With that, we can find good candidates (with pure perf) to dive into.

For the kernel benchmark in this PR, I can let the benchmark print latency and bandwidth information as well. This way, we can easily tune configs manually and see the impact on the latency/bandwidth of the kernel.

Reply:
imo we should ideally be able to have some flexibility in how we call the kernel and with how many arguments.

For example, ideally, the code would look something like
```
def call(args):
    arg0_1, = args
    args.clear()
    with torch.cuda._DeviceGuard(0):
        torch.cuda.set_device(0) # no-op to ensure context
        buf0 = empty_strided((2, 3), (3, 1), device='cuda', dtype=torch.float32)
        stream0 = get_cuda_stream(0)
        triton__0.run(arg0_1, buf0, 6, grid=grid(6), stream=stream0)
        del arg0_1
        return (buf0, )

def get_args():
      arg0_1 = rand_strided((2, 3), (3, 1), device='cuda:0', dtype=torch.float32)
      return (arg0_1,)

if __name__ == "__main__":
    from torch._dynamo.testing import rand_strided
    from torch._inductor.utils import print_performance
    args = get_args()
    print_performance(lambda: call(list(args)))
```

And then future work would be able to do something like
```
for kernel in kernel_dataset:
    # do something with kernel
```

Reply:
So given a kernel, its argument list is already fixed. Currently the generated benchmark code looks like:

```
if __name__ == "__main__":
    from torch._C import _cuda_getCurrentRawStream as get_cuda_stream
    from torch._dynamo.testing import rand_strided
    from torch._inductor.utils import print_performance
    import torch
    from torch._inductor.triton_ops.autotune import grid

    arg0_1 = rand_strided((2, 3), (3, 1), device='cuda:0', dtype=torch.float32)
    buf0 = rand_strided((2, 3), (3, 1), device='cuda:0', dtype=torch.float32)
    stream0 = get_cuda_stream(0)
    print_performance(lambda: triton_.run(arg0_1, buf0, 6, grid=grid(6), stream=stream0) or True)
    ```

I can add the missing DeviceGuard. One more difference I see in your example is, you treat arg0_1 as kernel input and buf0 as kernel output (although it's address is passed in as kernel argument). But in my example, they are alls randomly generated arguments. I think your way is better. The question I have is, do we have a good way to figure out which kernel arguments is actually for output?

Reply:
@shunting314 I'm saying that I think the main thing that's missing from this PR is the ability to pull out the kernel from the dumped file and have control over how often I'm calling it.

Like, if you import the kernel file,  you'll have access to a call function, but we don't have a way of calling that function since we don't have programmatic access to the arguments. Our only way of calling the function is to call `__main__`.

Reply:
ah got it.  I'll organized code into
- call function
- get_args function
- use the above 2 functions in `__main__`